test/csuite/wt5165_cursor_copy_debug/main.c

```diff
@@ -0,0 +1,253 @@
+/*-
+ * Public Domain 2014-2019 MongoDB, Inc.
+ * Public Domain 2008-2014 WiredTiger, Inc.
+ *
+ * This is free and unencumbered software released into the public domain.
+ *
+ * Anyone is free to copy, modify, publish, use, compile, sell, or
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "test_util.h"
+
+/*
+ * JIRA ticket reference: WT-5165 Test case description: Using the cursor copy debug mode, verify
+ * that keys and values returned by API calls are freed as appropriate (typically at the next API
+ * call boundary). Because it is not possible to portably tell if memory has been freed (as it may
+ * be reused by the allocator for something else), we instead make sure it does not have the
+ * previously set key or value in it. We do rely on the fact that after memory is freed, it remains
+ * a valid memory address, but is either on the memory free list, or has been reused.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+
+static const char *K1 = "key1";
+static const char *K2 = "key2222";
+static const char *K3 = "key333";
+
+static const char *V1 = "value1";
+static const char *V2 = "value2222";
+static const char *V3 = "value333";
+
+/*
+ * We expect that the memory was freed at some point by the WiredTiger API. However, that same
+ * memory may have been reused for another purpose. The best we can do is verify that either the
+ * memory is marked free by malloc, or if not, that it has been overwritten with some other value.
+ */
+#define ASSERT_FREE(mem, prev_mem)                       \
+    do {                                                 \
+        testutil_assert(strcmp(mem, prev_mem) != 0);     \
+        printf(" expect junk: %s\n", (const char *)mem); \
+    } while (0)
```

Comment:
This is potentially going to drop core if the memory is no longer part of the application's address space.

Reply:
Understood.  There's no great way to check if a memory address is valid before accessing.  Well, there's a clunky way that I used (use the address in a system call and see if `EACCESS` comes back).